ext/curl/interface.c

```diff
@@ -1185,12 +1190,119 @@ PHP_MINIT_FUNCTION(curl)
 		return FAILURE;
 	}
 
+	zend_class_entry ce;
+	INIT_CLASS_ENTRY(ce, "CurlHandle", class_CurlHandle_methods);
+	curl_ce = zend_register_internal_class(&ce);
+	curl_ce->ce_flags |= ZEND_ACC_FINAL;
+	curl_ce->create_object = curl_create_object;
+	curl_ce->serialize = zend_class_serialize_deny;
+	curl_ce->unserialize = zend_class_unserialize_deny;
+
+	memcpy(&curl_object_handlers, &std_object_handlers, sizeof(zend_object_handlers));
+	curl_object_handlers.offset = XtOffsetOf(php_curl, std);
+	curl_object_handlers.free_obj = curl_free_obj;
+	curl_object_handlers.get_gc = curl_get_gc;
+	curl_object_handlers.get_constructor = curl_get_constructor;
+	curl_object_handlers.clone_obj = curl_clone_obj;
+
+	curl_multi_register_class(class_CurlMultiHandle_methods);
+	curl_share_register_class(class_CurlShareHandle_methods);
 	curlfile_register_class();
 
 	return SUCCESS;
 }
 /* }}} */
 
+/* Curl class */
+
+static zend_object *curl_create_object(zend_class_entry *class_type) {
+	php_curl *intern = zend_object_alloc(sizeof(php_curl), class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+	intern->std.handlers = &curl_object_handlers;
+
+	return &intern->std;
+}
+
+static zend_function *curl_get_constructor(zend_object *object) {
+	zend_throw_error(NULL, "Cannot directly construct CurlHandle, use curl_init() instead");
+	return NULL;
+}
+
+static zend_object *curl_clone_obj(zend_object *object) {
+	php_curl *ch;
+	CURL *cp;
+	zval *postfields;
+	zend_object *clone_object;
+	php_curl *clone_ch;
+
+	clone_object = curl_create_object(curl_ce);
+	clone_ch = curl_from_obj(clone_object);
+	init_curl_handle(clone_ch);
+
+	ch = curl_from_obj(object);
+	cp = curl_easy_duphandle(ch->cp);
+	if (!cp) {
+		zend_throw_exception(NULL, "Cannot clone unconstructed Curl object", 0);
```

Comment:
```suggestion
		zend_throw_exception(NULL, "Failed to cloned CurlHandle", 0);
```

I don't think this is really limited to "unconstructed" curl objects, whatever that means.